### ISSUE TITLE
[R-Forge #2352] Columnwise set()?

### ISSUE THREAD

**Issue by arunsrinivasan** [Rdatatable/data.table#610]
Submitted by: Matt Dowle; Assigned to: Nobody; [R-Forge link](https://r-forge.r-project.org/tracker/index.php?func=detail&aid=2352&group_id=240&atid=978)

set() already works on data.frame.  Think the idea here might be a function that loops set() through each column to replace values.

Then add answer here :
http://stackoverflow.com/questions/12835942/fast-replacing-values-in-dataframe-in-r

**Comment by jangorecki** [comment]
I came up with something like this.
`.j` symbol used in `i` argument is iterated and replaced with columns from `j` argument.
I am not sure if we really want to have that in package.
```r
library(data.table)
set.seed(108)
d = data.table(a=rnorm(1000),b=rnorm(1000))
set.colwise = function(x, i, j, value) {
  replace_dot_j = function(e, j) {
    if (is.symbol(e) && identical(e, as.symbol(".j"))) return(j)
    if (is.call(e)) {
      if (e[[1L]] == ".j") e[[1L]] = j
      for (i in seq_along(e)[-1L]) if (!is.null(e[[i]])) e[[i]] = replace_dot_j(e[[i]], j)
    }
    e
  }
  for (jj in j) eval(substitute(
    set(x, .i, .j, value),
    list(
      .i=replace_dot_j(substitute(i), jj),
      .j=jj
    )
  ))
  invisible(x)
}
d
set.colwise(d, i = which(d[[.j]] < 0), j = c("a","b"), value = 0)
d
```

**Comment by MichaelChirico** [comment]
This is a wrapper for:

```
for (j in c('a', 'b')) set(d, which(d[[j]] < 0), j, 0)
```

?

**Comment by jangorecki** [comment]
@MichaelChirico yes, exactly what this issue is about. My vote would be for "won't fix", that is why comment...
> I am not sure if we really want to have that in package

there is always maintenance cost of such stuff, and there is NSE, which generally requires extra support.

**Comment by MichaelChirico** [comment]
Agreed, looks more like "convenience" function -- manual version is about the same to type and reads easier...

Might make sense if we decide to parallelize on the back end (especially since data.table assignment in parallel can be tough) but would want to see user demand for that before investing effort...

**Comment by jangorecki** [comment]
@MichaelChirico parallelise over columns would be safe, just need to push down all that to C.
In `nafill` branch #3341 there is clear example of parallel over columns `set` defined as:
```r
fsetnafill(x, fill, cols) set(x, j=cols, value=fill)
```

**Comment by jangorecki** [comment]
Closing this for now, if there will be demand for that we can re-open.
